Overhaul MapPreview rule loading

pchote opened the pull request:
This PR reworks the way MapPreview handles custom rules, with two major goals in mind:

1. Fix the atrocious performance issues with the mission browser. On my system, and I suspect most others, opening the mission browser will trigger unusable performance on the shellmap for 10+ seconds while the mission rules are cached in the "background" (doesn't feel very background to me).

2. Require maps to pass the lint checks before dedicated servers will allow them to be used. This is responding to the flood of crash reports and complaints from the competitive community caused by copying the previous generation RAGL maps into release-20210321 without verifying that they were actually compatible (they weren't).

The first point is achieved by replacing the general Ruleset parsing with focused parsing for just the world and player definitions, which cuts the parse time for mission/custom maps from ~250ms to ~25ms each. This is small enough to get away with removing the asynchronous code completely, and load the rules up-front at the cost of only 1-2s of additional startup time in RA (less in the other mods).

The previous (very limited) rule syntax checks in the lobby were not compatible with this new approach, so they are moved to the server using a new `LobbyInfo.GlobalSettings.MapStatus` field to relay the result back to the clients. This decoupled approach makes it easy to add additional and longer checks, so I scope creeped a solution to the second point while I was here.

If we can get this merged within a reasonable timeframe I expect to work on a second PR that would bring the "map not available on server" handling ~~under the same system~~ (edit: ended up taking a different approach), which would then naturally extend to supporting server-side map allow/deny lists to make @ubitux's life easier with managing the map pool on the competitive servers.

I strongly recommend reviewing by commit, which makes it a lot easier to understand the different parts of this.

pchote committed:
Remove direct access to MapPreview.Rules.

pchote committed:
Move LoadMaps after InitializeFonts.

This allows text to be displayed earlier in the loading screen.

pchote commented:
I realized while writing the PR description that I have forgotten to add a new lint rule to enforce the new World and Player inheritance restrictions.

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/MapPreviewLogic.cs:
```diff
@@ -171,31 +222,31 @@ public override void Tick()
 		}
 
-		void SetupWidgets(Widget parent, Func<MapPreview> getMap,
+		void SetupWidgets(Widget parent, Func<(MapPreview, ServerMapStatus)> getMap,
```
We usually define names for elements of tuples instead of using `Item1`.

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/ReplayBrowserLogic.cs:
```diff
@@ -93,7 +94,7 @@ public ReplayBrowserLogic(Widget widget, ModData modData, Action onExit, Action
 			{
 				{ "orderManager", null },
-				{ "getMap", (Func<MapPreview>)(() => map) },
+				{ "getMap", (Func<(MapPreview, ServerMapStatus)>)(() => (map, ServerMapStatus.Playable)) },
```
Same here.

pchote replied:
Done at the receiving side. I don't think there is any value in adding them here since the context is immediately obvious from the lambda, but can change this if you insist.

abcdefg30 replied:
Fair enough, we can keep it as is here.

pchote commented:
Updated. I have added the missing lint test and also bumped the orders protocol to avoid breakage from the alternative server implementation.

reaperrr commented:
> Fix the atrocious performance issues with the mission browser. On my system, and I suspect most others, opening the mission browser will trigger unusable performance on the shellmap for 10+ seconds while the mission rules are cached in the "background" (doesn't feel very background to me).

Hm, not on my system, opening the mission browser in RA is virtually instant here with no noticable lag on UI or shellmap afterwards (Win10, Ryzen 3600, 16GB DDR4-3200, SATA SSDs only).

Doesn't change that this overhaul is certainly justified, I just won't be able to test the performance difference since I can't reproduce the perf issue on bleed.

reaperrr reviewed (dismissed):
Didn't see any issues glancing over the code and mission/skirmish browsers still work fine for me

pchote commented:
Hmm, odd.

This is what it looks like for me on current bleed:

https://user-images.githubusercontent.com/167819/113574047-3c96e880-9613-11eb-8cc5-ddd4ebb7f755.mov

This only started relatively recently, not sure if it was due to the switch to .NET5, due to all the new missions we've merged over the last year, or something else. In any case, this PR puts an end to it.

pchote commented on OpenRA.Game/Map/MapPreview.cs:
```diff
@@ -469,17 +526,12 @@ void OnDownloadProgress(long total, long received, int percentage)
-							innerData.Status = MapStatus.DownloadError;
-						else
-						{
-							UpdateFromMap(package, mapInstallPackage, MapClassification.User, null, GridType);
```
This is needed for map installation to work!

pchote committed:
Rework MapPreview custom rule handling.

The previous asynchronous approach did not work particularly well,
leading to large janks when switching to custom maps or opening the
mission browser.

This commit introduces two key changes:

 * Rule loading for WorldActorInfo and PlayerActorInfo is made
   synchronous, in preparation for the next commit which will
   significantly optimize this path.
 * The full ruleset loading, which is required for map validation,
   is moved to the server-side and managed by a new ServerMapStatusCache.
   The previous syntax check is expanded to include the ability to run
   lint tests.

pchote committed:
Fix map installation.

pchote committed:
Optimize MapPreview rule loading.

pchote commented:
Updated:
* Moved `MapStatusCache` and `MapStatus` definitions from `OpenRA.Network` into `OpenRA.Server` and `OpenRA.Network.Session` respectively.
* Fixed compile error on "Optimize MapPreview rule loading." commit (it relied on changes that were introduced in the next commit)
* Fixed map installation (one regression from #18995 and a second introduced in this PR)

reaperrr commented:
Needs rebase.

pchote commented:
Rebased.

pchote commented:
Now needs a rebase/update for #19239.

pchote committed:
Enable dedicated server lint checks.

pchote commented:
Rebased.

teinarss commented on OpenRA.Game/Map/MapPreview.cs:
```diff
@@ -193,6 +219,26 @@ internal void SetMinimap(Sprite minimap)
+				innerData.NotificationDefinitions, innerData.SequenceDefinitions);
+		}
+
+		public bool HasCustomRules =>
```
Not used

teinarss reviewed (commented):
Also looks like `public enum MapRuleStatus` in MapPreview is not used.

pchote commented:
Updated.

teinarss commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/1e4f5783affbb5b8101763456c9f654a30cbdc29...d7cf9477eba791a2cc37e1041b5a471189d7637e
